File: go/internal/cli/robot_docs.go
package cli

import (
	"fmt"
	"strings"

	"github.com/spf13/cobra"
)

var robotDocsCmd = &cobra.Command{
	Use:     "robot-docs",
	Aliases: []string{"agent"},
	Short:   "Print in-tool guidance for automation agents",
	Long:    "Automation-focused guides for using ooxml without external documentation lookup.",
	Args:    cobra.NoArgs,
	RunE:    showHelp,
}

var robotDocsGuideCmd = &cobra.Command{
	Use:   "guide",
	Short: "Print a compact agent guide",
	Long:  "Print a paste-ready guide for fast PPTX, XLSX, and VBA automation with ooxml.",
	Args:  cobra.NoArgs,
	RunE: func(cmd *cobra.Command, args []string) error {
		guide := buildRobotDocsGuide()
		if GetGlobalConfig(cmd).Format == "json" {
			return writeGlobalJSON(cmd, guide)
		}
		return outputRobotDocsGuideText(cmd, guide)
	},
}

type robotDocsGuide struct {
	Tool       string             `json:"tool"`
	Version    string             `json:"version"`
	Principles []string           `json:"principles"`
	Sections   []robotDocsSection `json:"sections"`
	Warnings   []string           `json:"warnings"`
}

type robotDocsSection struct {
	Name     string   `json:"name"`
	Commands []string `json:"commands"`
}

func buildRobotDocsGuide() robotDocsGuide {
	return robotDocsGuide{
		Tool:    "ooxml",
		Version: Version,
		Principles: []string{
			"Use --json for agent-readable output. It is the shortcut for --format json.",
			"Inspect before mutating, mutate to --out, then validate with --strict.",
			"Prefer stable selectors such as slide numbers, sheet names, table names, and shape/table targets from prior JSON output.",
			"Reuse generated command fields such as readbackCommand, showCommand, cellsExtractCommand, rangesExportCommand, and validateCommand when present.",
			"Use --in-place only when the caller explicitly wants the input file modified.",
		},
		Sections: []robotDocsSection{
			{
				Name: "Discovery",
				Commands: []string{
					"ooxml capabilities --json",
					"ooxml capabilities --json --for shape",
					"ooxml robot-docs guide",
					"ooxml agent guide",
					"ooxml --json inspect <file>",
					"ooxml validate --strict <file>",
				},
			},
			{
				Name: "Preflight and release proof",
				Commands: []string{
					"ooxml --json doctor",
					"ooxml --json doctor health",
					"ooxml --json doctor capabilities",
					"ooxml doctor robot-docs",
					"make check-release-fast",
					"make check-release-slow",
					"make check-office-vba-schema",
					"make check-office-vba-com",
					`powershell.exe -NoProfile -ExecutionPolicy Bypass -File .\tools\windows-office-edit-smoke.ps1 -RepoRoot . -MutationParallelism 4 -RequireOpenXmlSdk -RunConformance -SkipOffice`,
					`powershell.exe -NoProfile -ExecutionPolicy Bypass -File .\tools\windows-office-edit-smoke.ps1 -RepoRoot . -MutationParallelism 4 -OfficeOracleTimeoutSeconds 120 -RequireOpenXmlSdk -RunConformance`,
					`powershell.exe -NoProfile -ExecutionPolicy Bypass -File .\tools\windows-office-vba-smoke.ps1 -RepoRoot . -RequireOpenXmlSdk -SkipOffice -EnableVbaObjectModelAccess`,
					`powershell.exe -NoProfile -ExecutionPolicy Bypass -File .\tools\windows-office-vba-smoke.ps1 -RepoRoot . -RequireOpenXmlSdk -EnableVbaObjectModelAccess -OfficeOracleTimeoutSeconds 120`,
				},
			},
			{
				Name: "PPTX read",
				Commands: []string{
					"ooxml --json pptx slides list deck.pptx",
					"ooxml --json pptx slides show deck.pptx --slide 1 --include-text --include-bounds",
					"ooxml --json pptx slides selectors deck.pptx --slide 1",
					"ooxml --json pptx shapes show deck.pptx --slide 1",
					"ooxml --json pptx charts list deck.pptx",
					"ooxml --json pptx charts show deck.pptx --slide 1 --chart chart:1",
					"ooxml --json pptx tables show deck.pptx --slide 1 --target table:1",
					"ooxml --json pptx layouts show deck.pptx --layout 1",
				},
			},
			{
				Name: "PPTX mutate",
				Commands: []string{
					"ooxml --json pptx clone-slide deck.pptx --slide 1 --out edited.pptx",
					"ooxml --json pptx slides show edited.pptx --slide 2 --include-text --include-bounds",
					"ooxml --json pptx replace text deck.pptx --slide 1 --target title --text NEW --out edited.pptx",
					"ooxml --json pptx replace text-occurrences deck.pptx --match-text \"Old Client\" --new-text \"New Client\" --expect-count 12 --dry-run",
					"ooxml --json pptx replace text-occurrences deck.pptx --match-text \"Old Client\" --new-text \"New Client\" --expect-count 12 --expect-plan-hash sha256:... --out edited.pptx",
					"ooxml --json pptx tables set-cell deck.pptx --slide 1 --target table:1 --row 1 --col 1 --text Value --out edited.pptx",
					"ooxml --json pptx tables update-from-xlsx deck.pptx --workbook workbook.xlsx --sheet Sheet1 --range A1:C5 --expect-source-range A1:C5 --slide 1 --target table:1 --out edited.pptx",
					"ooxml --json pptx charts update-data deck.pptx --slide 1 --chart chart:1 --series 1 --values-json '[\"150\",\"175\",\"210\"]' --categories-json '[\"North\",\"South\",\"West\"]' --expect-point-count 3 --expect-values-hash sha256:... --out edited.pptx",
					"ooxml --json xlsx tables show workbook.xlsx --table Sales",
					"ooxml --json pptx place table-from-xlsx deck.pptx --workbook workbook.xlsx --table Sales --expect-source-range A1:C5 --slide 1 --x 0 --y 0 --cx 4000000 --out edited.pptx",
					"ooxml --json pptx place image deck.pptx --slide 1 --image hero.png --x 0 --y 0 --cx 4000000 --cy 2250000 --fit-mode cover --out edited.pptx",
					"ooxml --json pptx replace images deck.pptx --slide 2 --target shape:2 --image hero.png --fit-mode contain --out edited.pptx",
					"ooxml --json pptx shapes set-bounds deck.pptx --slide 2 --target body --bounds 914400,914400,7315200,3657600 --out edited.pptx",
					"ooxml --json pptx xlsx-bindings plan deck.pptx --workbook workbook.xlsx --table DeckBindings",
					"ooxml --json pptx xlsx-bindings apply deck.pptx --workbook workbook.xlsx --table DeckBindings --out edited.pptx",
					"ooxml --json pptx xlsx-bindings apply deck.pptx --workbook workbook.xlsx --table DeckImageBindings --dry-run",
					"ooxml --json pptx xlsx-bindings apply deck.pptx --workbook workbook.xlsx --table DeckBoundsBindings --out edited.pptx",
					"ooxml pptx render edited.pptx --out render-check",
					"ooxml validate --strict edited.pptx",
				},
			},
			{
				Name: "XLSX read",
				Commands: []string{
					"ooxml --json xlsx sheets list workbook.xlsx",
					"ooxml --json xlsx sheets show workbook.xlsx --sheet Sheet1",
					"ooxml --json xlsx cells extract workbook.xlsx --sheet Sheet1 --range A1:C10",
					"ooxml --json xlsx ranges export workbook.xlsx --sheet Sheet1 --range A1:C10 --include-types",
					"ooxml --json xlsx ranges set-format workbook.xlsx --sheet Sheet1 --range B2:B20 --preset currency --out edited.xlsx",
					"ooxml --json xlsx tables list workbook.xlsx --sheet Sheet1",
					"ooxml --json xlsx tables show workbook.xlsx --table Sales",
					"ooxml --json xlsx charts list workbook.xlsx",
					"ooxml --json xlsx charts show workbook.xlsx --chart chart:1",
					"ooxml --json xlsx pivots list workbook.xlsx",
					"ooxml --json xlsx pivots show workbook.xlsx --pivot pivot:1",
					"ooxml --json xlsx names list workbook.xlsx",
					"ooxml --json xlsx names show workbook.xlsx --name SalesData",
				},
			},
			{
				Name: "XLSX mutate",
				Commands: []string{
					"ooxml --json xlsx cells set workbook.xlsx --sheet Sheet1 --cell B2 --value '42' --out edited.xlsx",
					"ooxml --json xlsx ranges set workbook.xlsx --sheet Sheet1 --anchor A1 --values '[[\"A\",\"B\"],[1,2]]' --out edited.xlsx",
					"ooxml --json xlsx ranges set-format workbook.xlsx --sheet Sheet1 --range B2:B20 --preset currency --out edited.xlsx",
					"ooxml --json xlsx sheets rename workbook.xlsx --sheet OldName --name NewName --out edited.xlsx",
					"ooxml --json xlsx names add workbook.xlsx --name SalesData --sheet Sheet1 --range A1:C10 --out edited.xlsx",
					"ooxml --json xlsx names update edited.xlsx --name SalesData --sheet Sheet1 --range A1:D10 --expect-ref \"'Sheet1'!\\$A\\$1:\\$C\\$10\" --out edited.xlsx",
					"ooxml --json xlsx names delete edited.xlsx --name SalesData --expect-ref \"'Sheet1'!\\$A\\$1:\\$D\\$10\" --out edited.xlsx",
					"ooxml --json xlsx charts update-source workbook.xlsx --chart chart:1 --series 1 --role values --source-sheet Sheet1 --source-range '$B$2:$B$20' --expect-source-range '$B$2:$B$10' --out edited.xlsx",
					"ooxml --json xlsx tables append-rows workbook.xlsx --table Sales --values '[[\"Q1\",100]]' --out edited.xlsx",
					"ooxml validate --strict edited.xlsx",
				},
			},
			{
				Name: "VBA project and module operations",
				Commands: []string{
					"ooxml --json vba inspect workbook.xlsm",
					"ooxml --json vba create workbook.xlsm --family xlsx --source macros/Module1.bas --source macros/Worker.cls --extract-bin vbaProject.bin --enable-vba-object-model-access --force",
					"ooxml --json vba create deck.pptm --family pptx --source macros/Module1.bas --force",
					"ooxml --json vba extract-bin workbook.xlsm --out vbaProject.bin",
					"ooxml --json vba list workbook.xlsm",
					"ooxml --json vba extract workbook.xlsm --out-dir macros",
					"ooxml --json vba replace-module workbook.xlsm --module Module1 --source macros/Module1.bas --expect-sha256 <sha256-from-list> --out edited.xlsm",
					"ooxml --json vba add-module source-only.xlsm --source macros/NewModule.bas --expect-module-count 2 --out added-source-only.xlsm",
					"ooxml --json vba remove-module source-only-edited.xlsm --module Module1 --expect-sha256 <sha256-from-list> --out removed-module.xlsm",
					"ooxml --json vba attach target.xlsx --bin office-authored-vbaProject.bin --out target-with-vba.xlsm",
					"ooxml --json vba inspect target-with-vba.xlsm",
					"ooxml validate --strict target-with-vba.xlsm",
					"ooxml --json vba office-check target-with-vba.xlsm",
					"ooxml --json vba remove target.xlsm --out target-no-vba.xlsx",
					"ooxml --json vba inspect target-no-vba.xlsx",
				},
			},
		},
		Warnings: []string{
			"ooxml vba create can create fresh Office-authored XLSM/PPTM files from .bas/.cls sources on Windows desktop Office; other VBA commands inspect/extract source modules, replace existing modules, and add/remove modules only for synthetic/source-only projects. Real Office-shaped module-set changes are refused, so use vba create or an Office-authored vbaProject.bin plus attach for those edits. vba office-check provides local LibreOffice/soffice open-check evidence, but it does not execute or compile macros and is not Microsoft Office proof.",
			"After PPTX edits, validate and render before treating the file as user-ready.",
			"`pptx replace text-occurrences` updates slide-visible shape/table text only; notes, masters, layouts, charts, comments, and split-run phrase matches need explicit follow-up commands.",
			"After XLSX edits, validate and inspect the touched sheet/table before handing off.",
		},
	}
}

func outputRobotDocsGuideText(cmd *cobra.Command, guide robotDocsGuide) error {
	var b strings.Builder
	fmt.Fprintf(&b, "ooxml agent guide\n")
	fmt.Fprintf(&b, "Version: %s\n\n", guide.Version)

	fmt.Fprintf(&b, "Principles:\n")
	for _, principle := range guide.Principles {
		fmt.Fprintf(&b, "  - %s\n", principle)
	}

	for _, section := range guide.Sections {
		fmt.Fprintf(&b, "\n%s:\n", section.Name)
		for _, command := range section.Commands {
			fmt.Fprintf(&b, "  %s\n", command)
		}
	}

	fmt.Fprintf(&b, "\nWarnings:\n")
	for _, warning := range guide.Warnings {
		fmt.Fprintf(&b, "  - %s\n", warning)
	}

	return writeGlobalOutput(cmd, []byte(b.String()))
}

func init() {
	robotDocsCmd.AddCommand(robotDocsGuideCmd)
	rootCmd.AddCommand(robotDocsCmd)
}

File: src/robot_docs.rs
use serde_json::{Value, json};

use crate::agent_aliases::{
    CAPABILITY_COMMAND_FAMILY_FILTERS, CAPABILITY_OBJECT_KINDS, capability_filter_alias_strings,
};
use crate::cli_dispatch::{DispatchBody, DispatchOutput};
use crate::{CliError, CliResult, EXIT_SUCCESS, GlobalFlags, has_flag, reject_unknown_flags};

pub(crate) fn robot_docs(flags: &GlobalFlags, args: &[String]) -> CliResult<DispatchOutput> {
    match args {
        [sub, rest @ ..] if sub == "guide" => guide(flags, rest),
        [] => guide(flags, &[]),
        _ => Err(CliError::invalid_args(format!(
            "unsupported Rust-port contract command: robot-docs {}",
            args.join(" ")
        ))),
    }
}

pub(crate) fn agent_alias(flags: &GlobalFlags, args: &[String]) -> CliResult<DispatchOutput> {
    match args {
        [sub, rest @ ..] if sub == "guide" => guide(flags, rest),
        [sub, rest @ ..] if sub == "triage" => Ok(DispatchOutput {
            body: DispatchBody::Json(crate::agent_triage::agent_triage(rest)?),
            exit_code: EXIT_SUCCESS,
        }),
        _ => Err(CliError::invalid_args(format!(
            "unsupported Rust-port contract command: agent {}",
            args.join(" ")
        ))),
    }
}

fn guide(flags: &GlobalFlags, args: &[String]) -> CliResult<DispatchOutput> {
    reject_unknown_flags(args, &["--format"], &["--json"])?;
    let value = guide_json();
    if wants_json(flags, args) {
        Ok(DispatchOutput {
            body: DispatchBody::Json(value),
            exit_code: EXIT_SUCCESS,
        })
    } else {
        Ok(DispatchOutput {
            body: DispatchBody::Text(guide_text(&value)),
            exit_code: EXIT_SUCCESS,
        })
    }
}

fn wants_json(flags: &GlobalFlags, args: &[String]) -> bool {
    flags.json
        || has_flag(args, "--json")
        || args
            .windows(2)
            .any(|pair| (pair[0] == "--format" || pair[0] == "-f") && pair[1] == "json")
        || args
            .iter()
            .any(|arg| arg == "--format=json" || arg == "-f=json")
}

fn guide_json() -> Value {
    json!({
        "tool": "ooxml",
        "version": env!("CARGO_PKG_VERSION"),
        "principles": [
            "Use --json for machine-readable output whenever a command supports it.",
            "Inspect before mutating and keep readback commands from mutation output.",
            "Prefer stable selectors, handles, hashes, and generated command fields over positional guesses.",
            "Reuse generated command fields for validation and readback.",
            "Use --in-place only when the user explicitly asks to overwrite the input."
        ],
        "sections": [
            {
                "name": "Discovery",
                "commands": [
                    "ooxml --json capabilities",
                    "ooxml --json capabilities --for slides",
                    "ooxml --json capabilities --for conditional-formats",
                    "ooxml --json capabilities --for modules",
                    "ooxml agent-triage",
                    "ooxml robot-docs guide",
                    "ooxml --json inspect <file>",
                    "ooxml --json find <query> <file>",
                    "ooxml --json doctor health"
                ],
                "filters": CAPABILITY_COMMAND_FAMILY_FILTERS
                    .iter()
                    .copied()
                    .chain(CAPABILITY_OBJECT_KINDS.iter().copied())
                    .collect::<Vec<_>>(),
                "filterAliases": capability_filter_alias_strings()
            },
            {
                "name": "Preflight and release proof",
                "commands": [
                    "ooxml validate --strict <file>",
                    "ooxml --json doctor",
                    "ooxml --json doctor capabilities",
                    "ooxml --json conformance coverage",
                    "ooxml --json conformance check <file>",
                    "ooxml --json repair normalize <file> --out <normalized-file>"
                ]
            },
            {
                "name": "PPTX read",
                "commands": [
                    "ooxml --json pptx slides list <file>",
                    "ooxml --json pptx shapes show <file> --slide <n>",
                    "ooxml --json pptx tables show <file> --slide <n>",
                    "ooxml --json pptx charts list <file>",
                    "ooxml --json pptx extract text <file>",
                    "ooxml --json pptx extract notes <file>"
                ]
            },
            {
                "name": "PPTX mutate",
                "commands": [
                    "ooxml --json pptx replace text-occurrences <file> --match-text <old> --new-text <new> --out <file>",
                    "ooxml --json pptx tables set-cell <file> --slide <n> --target <selector> --row <n> --col <n> --text <text> --out <file>",
                    "ooxml --json pptx notes set <file> --slide <n> --text <text> --out <file>",
                    "ooxml --json pptx shapes set-bounds <file> --slide <n> --target <selector> --bounds <json> --out <file>"
                ]
            },
            {
                "name": "XLSX read",
                "commands": [
                    "ooxml --json xlsx sheets list <file>",
                    "ooxml --json xlsx cells extract <file>",
                    "ooxml --json xlsx ranges export <file> --sheet <sheet> --range <range>",
                    "ooxml --json xlsx names list <file>",
                    "ooxml --json xlsx tables list <file>",
                    "ooxml --json xlsx charts list <file>"
                ]
            },
            {
                "name": "XLSX mutate",
                "commands": [
                    "ooxml --json xlsx cells set <file> --sheet <sheet> --cell <cell> --value <value> --out <file>",
                    "ooxml --json xlsx ranges set <file> --sheet <sheet> --range <range> --values <json> --out <file>",
                    "ooxml --json xlsx names add <file> --name <name> --ref <ref> --out <file>",
                    "ooxml --json xlsx tables append-rows <file> --table <table> --values <json> --out <file>"
                ]
            },
            {
                "name": "DOCX read",
                "commands": [
                    "ooxml --json docx text <file>",
                    "ooxml --json docx blocks show <file>",
                    "ooxml --json docx tables show <file>",
                    "ooxml --json docx comments list <file>",
                    "ooxml --json docx images list <file>"
                ]
            },
            {
                "name": "DOCX mutate",
                "commands": [
                    "ooxml --json docx paragraphs append <file> --text <text> --out <file>",
                    "ooxml --json docx paragraphs set <file> --index <n> --text <text> --out <file>",
                    "ooxml --json docx tables create <file> --values <json-matrix> --out <file>",
                    "ooxml --json docx tables set-cell <file> --table <n> --row <n> --col <n> --expect-hash <hash> --text <text> --out <file>",
                    "ooxml --json docx comments add <file> --anchor-block <n> --author <name> --text <text> --out <file>"
                ]
            },
            {
                "name": "VBA package operations",
                "commands": [
                    "ooxml --json vba inspect <file>",
                    "ooxml --json vba build-bin --family xlsx|pptx|docx --source <module.bas> --out <vbaProject.bin>",
                    "ooxml --json vba create <file.xlsx|file.pptx|file.docx> --pure --source <module.bas> --out <file.xlsm|file.pptm|file.docm>",
                    "ooxml --json vba list <file>",
                    "ooxml --json vba extract <file> --out-dir <dir>",
                    "ooxml --json vba attach <file> --bin <vbaProject.bin> --out <file>",
                    "ooxml --json vba remove <file> --out <file>",
                    "ooxml --json vba run-smoke [file.xlsm] --smoke-mode Standard|Class --out-dir <proof-dir>",
                    "ooxml --json convert xlsm-to-xlsx <file.xlsm> --out <file.xlsx>"
                ]
            }
        ],
        "warnings": [
            "Do not run mutation commands without an explicit --out, --dry-run, or user-approved --in-place path.",
            "Do not assume legacy-only commands exist in the Rust CLI; use capabilities before invoking a command family.",
            "Use vba office-check for macro package open proof; on Windows it prefers Microsoft Office COM, but it does not execute macros.",
            "Use vba run-smoke only when the user explicitly asks for macro execution proof; it runs a harmless Excel VBA macro through desktop Office COM."
        ]
    })
}

fn guide_text(value: &Value) -> String {
    let mut out = String::from("OOXML robot guide\n\nPrinciples:\n");
    if let Some(principles) = value["principles"].as_array() {
        for item in principles {
            out.push_str("- ");
            out.push_str(item.as_str().unwrap_or_default());
            out.push('\n');
        }
    }
    if let Some(sections) = value["sections"].as_array() {
        for section in sections {
            out.push('\n');
            out.push_str(section["name"].as_str().unwrap_or_default());
            out.push_str(":\n");
            if let Some(commands) = section["commands"].as_array() {
                for command in commands {
                    out.push_str("- ");
                    out.push_str(command.as_str().unwrap_or_default());
                    out.push('\n');
                }
            }
            if let Some(filters) = section["filters"].as_array()
                && !filters.is_empty()
            {
                out.push_str("  Filters: ");
                out.push_str(
                    &filters
                        .iter()
                        .filter_map(Value::as_str)
                        .collect::<Vec<_>>()
                        .join(", "),
                );
                out.push('\n');
            }
            if let Some(aliases) = section["filterAliases"].as_array()
                && !aliases.is_empty()
            {
                out.push_str("  Filter aliases: ");
                out.push_str(
                    &aliases
                        .iter()
                        .filter_map(Value::as_str)
                        .collect::<Vec<_>>()
                        .join(", "),
                );
                out.push('\n');
            }
        }
    }
    if let Some(warnings) = value["warnings"].as_array() {
        out.push_str("\nWarnings:\n");
        for warning in warnings {
            out.push_str("- ");
            out.push_str(warning.as_str().unwrap_or_default());
            out.push('\n');
        }
    }
    out
}
